File: src/main/java/io/github/faecraft/heartshapedbox/body/BodyPartProvider.java
package io.github.faecraft.heartshapedbox.body;

import net.minecraft.nbt.CompoundTag;
import net.minecraft.util.Identifier;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Optional;

/**
 * Duck, applied in BodyPartDuck mixin
 */
public interface BodyPartProvider {
    HashMap<Identifier, AbstractBodyPart> getPartsMap();
    
    ArrayList<AbstractBodyPart> getParts();
    
    CompoundTag writeToTag();
    
    void readFromTag(CompoundTag tag);
    
    Optional<AbstractBodyPart> maybeGet(Identifier identifier);
    
    @Nullable AbstractBodyPart getOrNull(Identifier identifier);
    
    AbstractBodyPart getOrThrow(Identifier identifier);
}

File: src/main/kotlin/io/github/faecraft/heartshapedbox/body/BodyPartProvider.kt
package io.github.faecraft.heartshapedbox.body

import net.minecraft.nbt.CompoundTag
import net.minecraft.util.Identifier
import java.util.*

/**
 * Duck, applied in BodyPartDuck mixin
 */
public interface BodyPartProvider {
    public val partsMap: HashMap<Identifier, AbstractBodyPart>
    public val parts: ArrayList<AbstractBodyPart>
    public fun writeToTag(): CompoundTag
    public fun readFromTag(tag: CompoundTag)
    public fun maybeGet(identifier: Identifier): Optional<AbstractBodyPart>
    public fun getOrNull(identifier: Identifier): AbstractBodyPart?
    public fun getOrThrow(identifier: Identifier): AbstractBodyPart
}
